### PR TITLE
[feature] perform captive portal authentication by submitting a form

### LINKED ISSUE #60: [wifi-login-pages] Perform captive portal authentication
Captive portal authentication must be done by submitting a form (we can't use ajax unfortunately).

A form like the following one:

```html  
<form name="userauth" method="post" class="hidden"
      id="captive-portal-form"
      action="http://cp.domain.com/index.php?etc">
    <input type="hidden" name="auth_user">
    <input type="hidden" name="auth_pass">
    <input type="hidden" name="zone" value="zone_name">
    <input type="hidden" name="redirurl" value="https://wifi-login-app/status">
    <input type="hidden" name="accept" value="accept">
    <input type="submit" value="Login">
</form>
```

We should have a way to define fields dynamically in the configuration, eg:

```yaml
captive_portal_form:
    name: user_auth
    method: post
    action: http://cp.domain.com/index.php?etc
    fields:
        username: auth_user
        password: auth_pass
    additional_fields:
        - zone: zone_name
        - redirurl: https://wifi-login-app/status
        - accept: accept
```

The username and password fields have to be automatically filled after login & reauthentication and then the form should be submitted automatically.
~~The user should not see the status page at all, it will be redirected to it by the captive portal.~~

**EDIT**: there's an alternative, using an iframe, I just tested this and it works, the captive portal form is included with an iframe (hidden with css) and submitted there, we can use the `load` event of the iframe to find out the result from the captive portal and only then show the status page or if there was any error, like when the maximum daily bandwidth has been reached, the error will be shown to the user and the operation canceled.

This needs test + docs.

### LINKED ISSUE #61: [wifi-login-pages] Status page should not use authorize API from openwisp-radius
The authorize API endpoint from openwisp-radius should not be used at all.

### LINKED ISSUE #66: [wifi-login-pages] set mac address of the user to sessionStorage
The login page should be changed to get the `calling_station_id` (the mac address of the user) querystring parameter and save it to `sessionStorage`.

### LINKED ISSUE #67: [wifi-login-pages] Perform captive portal logout
Depends on #66.

Similar to #60 but for the logout.

This will affect the current logout page.

Before showing the logout button:

- check if there are active sessions for the user
  we need to add a server route to get the  user radius accounting, filter by is_open=true, and mac address (`calledStationid`) from #66 if present;
- if there is, show the logout button
- the logout button should do what it does now but instead of redirecting to the login page it should perform the captive portal logout by posting a form like the following:

```
<form action="<CAPTIVE_PORTAL_URL>" method="<METHOD(default to post)" id="logout" class="hidden">
<input name="logout_id" type="hidden" value="" id="session-id">
<input type="submit">
</form>
```

This form can be included in an iframe like in #65.

- after the form has been submitted successfully show the login page
- if there's no active session, issue a toast notification and redirect to log-in

### LINKED ISSUE #68: [wifi-login-pages] Login with iframe after signup
Ensure the same behavior of #60 is used but after successful signup.

### PR DESCRIPTION
**Pull request by Vivekrajput20**
Closes #60, closes #61, closes #66, closes #67 , closes #68 
Depends on [openwisp-radius#130](https://github.com/openwisp/openwisp-radius/pull/130)

### CODE REVIEW

File: client/components/status/status.js [hunk 1]
```diff
@@ -9,47 +9,123 @@ import { Link } from 'react-router-dom';
...
+      if (reply && this.ifameRef.current.contentDocument.title.indexOf("404") < 0) {
+        logout(cookies, orgSlug);
+        toast.error(reply, {
+          onOpen: () => toast.dismiss(mainToastId),
+        });
+      }
+    }
+  }
+
+
```
Comment (nemesisdesign) [inline client/components/status/status.js:86]: looks like there's one extra blank line

File: client/components/status/status.js [hunk 2]
```diff
@@ -108,6 +184,23 @@ export default class Status extends React.Component {
             </div>
           </div>
         </div>
+        {captivePortalForm && (window.top === window.self) &&
+          <>
+            <form ref={this.formRef} method={captivePortalForm.method || ''}
```
Comment (nemesisdesign) [inline client/components/status/status.js:189]: why does this default to empty string?

File: internals/generators/organization/config.yml.hbs [hunk 3]
```diff
@@ -323,6 +324,20 @@ client:
...
+      method: post
+      action: http://localhost/
+      fields:
+        username: auth_user
+        password: auth_pass
+      additional_fields:
+        - name: zone
+          value: zone_name
+        - name: redirurl
+          value: http://0.0.0.0:8080/{{slug}}/status
```
Comment (nemesisdesign) [inline internals/generators/organization/config.yml.hbs:337]: isn't `localhost` or `127.0.0.1` more common? It may be less confusing for most users.

File: client/components/status/status.js [hunk 4]
```diff
@@ -108,6 +184,23 @@ export default class Status extends React.Component {
...
+              target="owisp-auth-iframe"
+              className="owisp-auth-hidden"
+            >
+              <input readOnly type="text" name={captivePortalForm.fields.username || ''} value={username} />
+              <input readOnly type="text" name={captivePortalForm.fields.password || ''} value={password} />
+              {captivePortalForm.additional_fields.length > 0 ? captivePortalForm.additional_fields.map(field =>
+                <input readOnly type="text" name={field.name} value={field.value} />
+              ) : null}
+            </form>
+            <iframe onLoad={this.handleIframe} ref={this.ifameRef} name="owisp-auth-iframe" className="owisp-auth-hidden" title='owisp-auth-iframe' />
```
Comment (nemesisdesign) [inline client/components/status/status.js:201]: you're using single quotes for `owisp-auth-iframe` while double quotes for all other attrs

File: client/components/status/status.js [hunk 5]
```diff
@@ -108,6 +183,23 @@ export default class Status extends React.Component {
...
+              target="owisp-auth-iframe"
+              className="owisp-auth-hidden"
+            >
+              <input readOnly type="text" name={captivePortalForm.fields.username || ""} value={username} />
+              <input readOnly type="text" name={captivePortalForm.fields.password || ""} value={password} />
+              {captivePortalForm.additional_fields.length > 0 ? captivePortalForm.additional_fields.map(field =>
+                <input readOnly type="text" name={field.name} value={field.value} />
+              ) : null}
+            </form>
+            <iframe onLoad={this.handleIframe} ref={this.ifameRef} name="owisp-auth-iframe" className="owisp-auth-hidden" title="owisp-auth-iframe" />
```
Comment (nemesisdesign) [inline client/components/status/status.js:200]: can we indent this a bit more to avoid the line being too long?

File: client/components/status/status.js [hunk 6]
```diff
@@ -108,6 +183,23 @@ export default class Status extends React.Component {
             </div>
           </div>
         </div>
+        {captivePortalForm && (window.top === window.self) &&
+          <>
```
Comment (nemesisdesign) [inline client/components/status/status.js:187]: what does this mean? I should refresh my react fu, maybe you can help me?

File: client/components/status/status.js [hunk 7]
```diff
@@ -108,6 +183,23 @@ export default class Status extends React.Component {
...
+              className="owisp-auth-hidden"
+            >
+              <input readOnly type="text" name={captivePortalForm.fields.username || ""} value={username} />
+              <input readOnly type="text" name={captivePortalForm.fields.password || ""} value={password} />
+              {captivePortalForm.additional_fields.length > 0 ? captivePortalForm.additional_fields.map(field =>
+                <input readOnly type="text" name={field.name} value={field.value} />
+              ) : null}
+            </form>
+            <iframe onLoad={this.handleIframe} ref={this.ifameRef} name="owisp-auth-iframe" className="owisp-auth-hidden" title="owisp-auth-iframe" />
+          </>
```
Comment (nemesisdesign) [inline client/components/status/status.js:201]: also here you can help me out refreshing my react fu, why is this needed?

File: internals/generators/organization/config.yml.hbs [hunk 8]
```diff
@@ -323,6 +324,20 @@ client:
         text:
           en: "Go back to homepage"
 
+    captive_portal_form:
+      method: post
+      action: http://localhost/
```
[the remark this answers is not in the document]
Comment (Vivekrajput20) [reply internals/generators/organization/config.yml.hbs:329]: Sure

File: client/components/status/status.js (hunk 6 above)
Comment (Vivekrajput20) [reply client/components/status/status.js:187]: It is possible that captive portal will redirect to the status page, in that case status page will go into a infinite loop where each IFrame will render a status page + a new Iframe, to break that I've added this check to execute this block only in root window.

File: client/components/status/status.js (hunk 7 above)
Comment (Vivekrajput20) [reply client/components/status/status.js:201]: ref is passed to get a reference to that element inside react class, I'm submitting the form in this iframe and have added this onLoad handler which checks for  reply param and 404 after the response is received from captive protal and loaded in the Iframe.

File: client/components/status/status.js (hunk 6 above)
Comment (nemesisdesign) [reply client/components/status/status.js:187]: can we add a comment?

File: client/components/status/status.js (hunk 7 above)
Comment (nemesisdesign) [reply client/components/status/status.js:201]: also here can we add a comment?

File: internals/generators/organization/config.yml.hbs (hunk 3 above)
Comment (Vivekrajput20) [reply internals/generators/organization/config.yml.hbs:337]: Currently it's passed as an array field and it's name can be different for different captive portals, so I don't have any way to identify this field and modify it's value on fronted app. We can ask for captive portal's root url and save it in yml file and then use that variable in all the places where relative url is required.

**Comment by Vivekrajput20** [comment]
@nemesisdesign  please test the lastest PR, I've updated login with saving macaddr in session and also added logout functionality.

File: client/components/status/status.js [hunk 9]
```diff
@@ -1,55 +1,178 @@
...
-        token,
-      }),
-    })
-      .then(response => {
-        if (response.data["control:Auth-Type"] !== "Accept") {
+    const url = getUserRadiusSessionsUrl(orgSlug);
+    const macaddr = cookies.get(`${orgSlug}_macaddr`);
+    try {
+      const response = await axios({
+        method: "post",
```
[the remark this answers is not in the document]
Comment (Vivekrajput20) [reply client/components/status/status.js:65]: @nemesisdesign  this is a request to proxy server and proxy server actually sends a GET request only, I'll convert it to GET request.

[the remark this answers is not in the document]
Comment (Vivekrajput20) [reply client/components/status/status.js]: This was a request to proxy server and proxy server was sending the GET request only. I've updated it to GET in the latest push.

**Comment by Vivekrajput20** [comment]
@nemesisdesign  It works in my local so I can't exactly figure out what's the bug, the configuration is fine and should have worked.
These are some of the things that can be causing this error.
-  I added `@babel/plugin-transform-runtime` package in this PR, so this PR needs npm install before proceeding.
- Frontend server starts at `0.0.0.0:8080/` and if the captive portal redirurl is `http://localhost:8080/default/status` and not `http://0.0.0.0:8080/default/status` this won't work as a document can access an Iframe's url only if domain is same. And 0.0.0.0 and localhost are considered different domains.

I've improved the PR with few more checks.

**Comment by Vivekrajput20** [comment]
@nemesisdesign that localhost is defined in default org's configuration ([here](https://github.com/openwisp/openwisp-wifi-login-pages/blob/4d1fa882cd3232a5ab59bc163f29703a48cc3054/org-configurations/default-configuration.yml#L301))
I think you are accessing captive portal for default organization that's why this request to local host is made. You will have to change the organization slug in the browser url.

**Comment by Vivekrajput20** [comment]
@nemesisdesign there was a small mistake in the order of declaration, I've fixed it and now sessions will not be empty.
I added a loader as well after the logout button is clicked.
Regarding the server side logging, I also felt the need of it while developing. I'm opening a separate issue for it. Will try to implement it when I get some time.

**Comment by nemesisdesign** [comment]
> @nemesisdesign there was a small mistake in the order of declaration, I've fixed it and now sessions will not be empty.
> I added a loader as well after the logout button is clicked.
> macaddr value is stored in a cookie, you can check the cookies of the site. It's only updated in iframe's onload when it gets a query param with macaddr and will not be updated by simply adding it to the browser url.

Ah ok I understand, there was a misunderstanding between us for the initial requirement.

The captive portal will redirect the user to the captive page with the `macaddr` querystring param in the URL, that's the way we should support. Is it clearer now?

> Regarding the server side logging, I also felt the need of it while developing. I'm opening a separate issue for it. Will try to implement it when I get some time.

Ok thanks.

**Review by nemesisdesign — COMMENTED** [review]
> @nemesisdesign there was a small mistake in the order of declaration, I've fixed it and now sessions will not be empty.
> I added a loader as well after the logout button is clicked.
> macaddr value is stored in a cookie, you can check the cookies of the site. It's only updated in iframe's onload when it gets a query param with macaddr and will not be updated by simply adding it to the browser url.

Ah ok I understand, there was a misunderstanding between us for the initial requirement.

The captive portal will redirect the user to the captive page with the `macaddr` querystring param in the URL, that's the way we should support. Is it clearer now?

> Regarding the server side logging, I also felt the need of it while developing. I'm opening a separate issue for it. Will try to implement it when I get some time.

Ok thanks.

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
Thanks for the update @Vivekrajput20,

I tried running the latest version of this branch but I get this error when I run `npm start`:

```
npm ERR! code ELIFECYCLE
npm ERR! errno 1
[0] npm ERR! openwisp-wifi-login-pages@1.0.0 server: `npm run setup && nodemon server/start.js`
[0] npm ERR! Exit status 1
npm ERR! 
[0] npm ERR! Failed at the openwisp-wifi-login-pages@1.0.0 server script.
[0] npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
[0] 
[0] npm ERR! A complete log of this run can be found in:
[0] npm ERR!     /home/nemesis/.npm/_logs/2020-09-06T22_37_36_382Z-debug.log
[0] npm run server exited with code 1
[1] No server url specified. Expecting server to run on http://localhost:3030
[1] ℹ ｢wds｣: Project is running at http://0.0.0.0:8080/
[1] ℹ ｢wds｣: webpack output is served from /
[1] ℹ ｢wds｣: Content not from webpack is served from /home/nemesis/Code/openwisp/openwisp-wifi-login-pages/public
[1] ℹ ｢wds｣: 404s will fallback to /index.html
[1] events.js:174
[1]       throw er; // Unhandled 'error' event
[1]       ^
[1] 
[1] Error: ENOSPC: System limit for number of file watchers reached, watch '/home/nemesis/Code/openwisp/openwisp-wifi-login-pages/public'
[1]     at FSWatcher.start (internal/fs/watchers.js:165:26)
[1]     at Object.watch (fs.js:1258:11)
[1]     at createFsWatchInstance (/home/nemesis/Code/openwisp/openwisp-wifi-login-pages/node_modules/chokidar/lib/nodefs-handler.js:38:15)
[1]     at setFsWatchListener (/home/nemesis/Code/openwisp/openwisp-wifi-login-pages/node_modules/chokidar/lib/nodefs-handler.js:81:15)
[1]     at FSWatcher.NodeFsHandler._watchWithNodeFs (/home/nemesis/Code/openwisp/openwisp-wifi-login-pages/node_modules/chokidar/lib/nodefs-handler.js:233:14)
[1]     at FSWatcher.NodeFsHandler._handleDir (/home/nemesis/Code/openwisp/openwisp-wifi-login-pages/node_modules/chokidar/lib/nodefs-handler.js:429:19)
[1]     at FSWatcher.<anonymous> (/home/nemesis/Code/openwisp/openwisp-wifi-login-pages/node_modules/chokidar/lib/nodefs-handler.js:477:19)
[1]     at FSWatcher.<anonymous> (/home/nemesis/Code/openwisp/openwisp-wifi-login-pages/node_modules/chokidar/lib/nodefs-handler.js:482:16)
[1]     at FSReqWrap.oncomplete (fs.js:154:5)
[1] Emitted 'error' event at:
[1]     at FSWatcher._handleError (/home/nemesis/Code/openwisp/openwisp-wifi-login-pages/node_modules/chokidar/index.js:260:10)
[1]     at createFsWatchInstance (/home/nemesis/Code/openwisp/openwisp-wifi-login-pages/node_modules/chokidar/lib/nodefs-handler.js:40:5)
[1]     at setFsWatchListener (/home/nemesis/Code/openwisp/openwisp-wifi-login-pages/node_modules/chokidar/lib/nodefs-handler.js:81:15)
[1]     [... lines matching original stack trace ...]
[1]     at FSReqWrap.oncomplete (fs.js:154:5)
[1] npm ERR! code ELIFECYCLE
[1] npm ERR! errno 1
[1] npm ERR! openwisp-wifi-login-pages@1.0.0 client: `npm run setup && webpack-dev-server --host 0.0.0.0 --mode development --config config/webpack.config.js`
[1] npm ERR! Exit status 1
```
Do you know what is going on? How can I fix this?

Also, I want to underline what I wrote in the previous comment:

> @nemesisdesign there was a small mistake in the order of declaration, I've fixed it and now sessions will not be empty.
> I added a loader as well after the logout button is clicked.
> macaddr value is stored in a cookie, you can check the cookies of the site. It's only updated in iframe's onload when it gets a query param with macaddr and will not be updated by simply adding it to the browser url.

There was a misunderstanding between us for the initial requirement.

The captive portal will redirect the user to the captive page with the `macaddr` querystring param in the URL, that's the way we should support. Is it clearer now?

### Tests

I think the solution is very near completion, the login works, the logout should work after your changes are fixed, the `macaddr` param issue should be easy to fix, I think this is a good time to fix the tests as well.

**Comment by Vivekrajput20** [comment]
@nemesisdesign  the fix for this issue is mentioned [here](https://github.com/gatsbyjs/gatsby/issues/11406#issuecomment-458769756).
I'll add support for query sting in the status url. And will also start fixing the tests.

File: client/components/status/status.js [hunk 10]
```diff
@@ -1,55 +1,178 @@
...
+    const {cookies, orgSlug} = this.props;
+    // to prevent recursive call in case redirect url is status page
+    if (window.top === window.self) {
+      const {sessions} = this.state;
+      const isValid = await this.validateToken();
+      if (isValid) {
+        const macaddr = cookies.get(`${orgSlug}_macaddr`);
+        if (macaddr) {
+          await this.getUserRadiusSessions();
+          if (sessions && sessions.length === 0) {
```
Comment (nemesisdesign) [inline client/components/status/status.js:49]: I think we should add a comment above this to explain we'll do the log in only if there's no other active session.
BTW I think we should use the mac address param (if not empty) here to filter the sessions, because otherwise users may not be able to log in with two different devices using the same account

**Review by nemesisdesign — COMMENTED** [review]
> @nemesisdesign the fix for this issue is mentioned [here](https://github.com/gatsbyjs/gatsby/issues/11406#issuecomment-458769756).

Thanks, that worked.

I tried again but the POST request to the captive portal logout is not firing, even though the code looks ok:

logout-broken

> I'll add support for query string in the status url. And will also start fixing the tests.

Ok, thanks!

**Comment by Vivekrajput20** [comment]
> I tried again but the POST request to the captive portal logout is not firing, even though the code looks ok
@nemesisdesign It must be firing logout request but as you have selected only XHR requests in the network that is why it's not showing. Since it's a HTML request it will be visible in All requests.

File: client/components/status/status.js (hunk 10 above)
Comment (Vivekrajput20) [reply client/components/status/status.js:49]: @nemesisdesign  Since mac address is stored in a cookie, it's already available to the proxy server and there was no need to send it from front-end. I've already added this filter to [proxy server's code](https://github.com/openwisp/openwisp-wifi-login-pages/blob/432e88b4c7913b129f5b30835535c965a86c3d47/server/controllers/user-radius-sessions-controller.js#L23)
And I'll add the comment explaining that.

File: client/components/status/status.js [hunk 11]
```diff
@@ -1,55 +1,191 @@
...
+    const {sessions} = this.state;
+    if (sessions.length > 0) {
+      if (this.logoutFormRef && this.logoutFormRef.current) {
+        this.logoutFormRef.current.submit();
+        setTimeout(() => {}, 0);
+      }
+    }
+    logout(cookies, orgSlug);
+    setLoading(false);
+    toast.success(logoutSuccess);
```
Comment (nemesisdesign) [inline client/components/status/status.js:108]: if I comment out these 3 lines (106, 107 and 108), the captive portal logout works.
I think these lines should be executed only once `this.logoutFormRef.current.submit()` has finished, otherwise it seems the submit is not triggered.
Can you add listener to the onload of the iframe or anything similar so we can execute these 3 lines only after the captive portal logout has completed?

Keep in mind the same should happen for the login as well, if the captive portal login has to be executed, the success message should be shown only once the captive portal has replied, otherwise users may try to visit another page and be redirected again to the captive page and then they'll be super confused.

With this solved we're ready to merge and move on! :+1: :clap:

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
Great progress Vivek, thank you!

I tested and debugged it again and found the reason for which the captive portal logout is not working, I left a comment below.

I also spotted this console warning:

Screenshot from 2020-09-09 21-30-11

**Review by nemesisdesign — APPROVED** [review]
It works! Thanks @Vivekrajput20 :+1: :clap: :tada:
